plugins/eos/gs-plugin-eos.c

```diff
@@ -267,6 +302,243 @@ read_icon_replacement_overrides (GHashTable *replacement_app_lookup)
+						 g_variant_dict_end (&options_dict),
+						 cancellable, error);
+}
+
+/* This method deals with the synchronization between the EOS updater's states
+ * (DBus service) and the OS upgrade's states (GsApp), in order to show the user
+ * what is happening and what they can do. */
+static EosUpdaterState
+sync_state_from_updater (GsPlugin *plugin)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	GsApp *app = get_os_upgrade (plugin);
+	EosUpdaterState state;
+	AsAppState previous_app_state = gs_app_get_state (app);
+	AsAppState current_app_state;
+	const gchar *version;
+
+	state = eos_updater_get_state (priv->updater_proxy);
+	g_debug ("EOS Updater state changed: %s", eos_updater_state_str [state]);
+
+	switch (state) {
+	case EOS_UPDATER_STATE_NONE:
+	case EOS_UPDATER_STATE_READY:
+		gs_app_set_state (app, AS_APP_STATE_UNKNOWN);
+		break;
+	case EOS_UPDATER_STATE_POLLING:
+		break;
+	case EOS_UPDATER_STATE_UPDATE_AVAILABLE: {
+		guint64 total_size = 0;
+		if (gs_app_get_state (app) == AS_APP_STATE_INSTALLING) {
+			/* when the OS upgrade is already being installed and the
+			 * updater reports an available update, (meaning we were
+			 * polling before), we should readily call fetch */
+			os_upgrade_force_fetch (priv->updater_proxy);
+			break;
+		}
+
+		total_size = eos_updater_get_download_size (priv->updater_proxy);
+		gs_app_set_size_download (app, total_size);
+
+		gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+
+		break;
+	}
+	case EOS_UPDATER_STATE_ERROR: {
+		const gchar *error_name;
+		const gchar *error_message;
+		g_autoptr(GError) local_error = NULL;
+
+		error_name = eos_updater_get_error_name (priv->updater_proxy);
+		error_message = eos_updater_get_error_message (priv->updater_proxy);
+		local_error = g_dbus_error_new_for_dbus_error (error_name, error_message);
+
+		g_debug ("Got OS upgrade error state with name '%s': %s",
+			 error_name, error_message);
+
+		gs_app_set_state_recover (app);
+
+		/* only set up an error to be shown to the user if the user had
+		 * manually started the upgrade, and if the error in question is not
+		 * originated by the user canceling the upgrade */
+		if (os_upgrade_get_download_by_user (app) &&
+		    g_strcmp0 (error_name, "com.endlessm.Updater.Error.Cancelled") != 0) {
+			g_autoptr(GsPluginEvent) event = gs_plugin_event_new ();
+			gs_utils_error_convert_gdbus (&local_error);
+			gs_plugin_event_set_app (event, app);
+			gs_plugin_event_set_error (event, local_error);
+			gs_plugin_event_add_flag (event, GS_PLUGIN_EVENT_FLAG_WARNING);
+			gs_plugin_report_event (plugin, event);
+		}
+
+		break;
+	}
+	case EOS_UPDATER_STATE_FETCHING: {
+		guint64 total_size = 0;
+		guint64 downloaded = 0;
+		gfloat progress = 0;
+
+		/* ensure the state transition to 'installing' is allowed */
+		if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING)
+			gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+		gs_app_set_state (app, AS_APP_STATE_INSTALLING);
```

Comment:
Is it because `INSTALLING` can only be reached from `AVAILABLE`? Can you refactor these into an `ensure_installing_state()` function?

Reply:
Yes. There are "safeguard" checks in the state transactions.
This logic is in only 2 places, so I thought I'd keep it like this. A more efficient force function would assign the state to `NULL` and then to whatever state we want, as that transaction is always allowed. But we do get a log about changing the state to `NULL` which can be confusing in the logs, so I preferred not to do that this time (I think I had done that before in other parts of the code) just for these cases + the one that keeps the `available` state.